cli/azd/pkg/project/ai_helper.go

```diff
@@ -0,0 +1,713 @@
+	envContainersClient *armmachinelearning.EnvironmentContainersClient,
+	envVersionsClient *armmachinelearning.EnvironmentVersionsClient,
+	modelContainersClient *armmachinelearning.ModelContainersClient,
+	modelVersionsClient *armmachinelearning.ModelVersionsClient,
+	endpointsClient *armmachinelearning.OnlineEndpointsClient,
+	deploymentsClient *armmachinelearning.OnlineDeploymentsClient,
+) AiHelper {
+	return &aiHelper{
+		env:                   env,
+		clock:                 clock,
+		pythonBridge:          pythonBridge,
+		workspacesClient:      workspacesClient,
+		envContainersClient:   envContainersClient,
+		envVersionsClient:     envVersionsClient,
+		modelContainersClient: modelContainersClient,
+		modelVersionsClient:   modelVersionsClient,
+		endpointsClient:       endpointsClient,
+		deploymentsClient:     deploymentsClient,
+	}
+}
+
+// RequiredExternalTools returns the required external tools for the AiHelper
+func (a *aiHelper) RequiredExternalTools(ctx context.Context) []tools.ExternalTool {
+	return a.pythonBridge.RequiredExternalTools(ctx)
+}
+
+// Initialize initializes the AiHelper
+func (a *aiHelper) Initialize(ctx context.Context) error {
+	if a.initialized {
+		return nil
+	}
+
+	if err := a.pythonBridge.Initialize(ctx); err != nil {
+		return err
+	}
+
+	a.initialized = true
+	return nil
+}
+
+// EnsureWorkspace ensures that the workspace exists
+func (a *aiHelper) EnsureWorkspace(
+	ctx context.Context,
+	scope *ai.Scope,
+) error {
+	workspaceName := scope.Workspace()
+
+	workspaceResponse, err := a.workspacesClient.Get(
+		ctx,
+		scope.ResourceGroup(),
+		workspaceName,
+		nil,
+	)
+	if err != nil {
+		return err
+	}
+
+	if *workspaceResponse.Workspace.Name != workspaceName {
+		return err
+	}
+
+	return nil
+}
+
+// CreateEnvironmentVersion creates a new environment version
+func (a *aiHelper) CreateEnvironmentVersion(
+	ctx context.Context,
+	scope *ai.Scope,
+	serviceConfig *ServiceConfig,
+	config *ai.ComponentConfig,
+) (*armmachinelearning.EnvironmentVersion, error) {
+	yamlFilePath := filepath.Join(serviceConfig.Path(), config.Path)
+	_, err := os.Stat(yamlFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	environmentName, err := config.Name.Envsubst(a.env.Getenv)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing environment name value: %w", err)
+	}
+
+	if environmentName == "" {
+		environmentName = fmt.Sprintf("%s-environment", serviceConfig.Name)
+	}
+
+	nextVersion := "1"
+	envContainerResponse, err := a.envContainersClient.Get(
+		ctx,
+		scope.ResourceGroup(),
+		scope.Workspace(),
+		environmentName,
+		nil,
+	)
+	if err == nil {
+		nextVersion = *envContainerResponse.Properties.NextVersion
+	}
+
+	environmentArgs := []string{
+		"-t", "environment",
+		"-s", scope.SubscriptionId(),
+		"-g", scope.ResourceGroup(),
+		"-w", scope.Workspace(),
+		"-f", yamlFilePath,
+		"--set", fmt.Sprintf("name=%s", environmentName),
+		"--set", fmt.Sprintf("version=%s", nextVersion),
```

Comment:
Just wondering - are we opening ourselves up to time of check/time of use bugs here?  What would happen if another version is deployed between when we saw this property and when we actually make the call to create the new version?  Do we need to be flowing an etag or something into the python helper so it can make a conditional request?

Reply:
This isn't something I want to get into at this point.  If multiple deployments were happening at the same time I'd let their data plane deal with that and report back the error.